### PR TITLE
initial changes to make document.getText() async

### PR DESCRIPTION
**Pull request by saidelike**
## Checklist

Base commit is f92f8890c3a21e6b74b470411de1c253236a2cc7

I created this draft PR in order to discuss the impact of trying to just change one function to be "async". This is not finished yet, but this gives an idea of the number of changes. 

It also makes the code less readable when dealing with "map", "flatMap", etc. 

There also edge cases where I am not sure how to deal with it yet. 

- [ ] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [ ] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/actions/BringMoveSwap.ts [hunk 1]
```diff
@@ -63,16 +63,16 @@ abstract class BringMoveSwap {
...
     destinations: Destination[],
-  ): ExtendedEdit[] {
+  ): Promise<ExtendedEdit[]> {
     const usedSources: Target[] = [];
     const results: ExtendedEdit[] = [];
     const shouldJoinSources =
       sources.length !== destinations.length && destinations.length === 1;
 
-    sources.forEach((source, i) => {
+    sources.forEach(async (source, i) => {
```
Comment (pokey) [inline packages/cursorless-engine/src/actions/BringMoveSwap.ts:75]:
```
for (let i = 0; i < source.length; i++) {
    const source = sources[i];
```

File: packages/cursorless-engine/src/processTargets/modifiers/InstanceStage.ts [hunk 2]
```diff
@@ -45,6 +47,7 @@ export class InstanceStage implements ModifierStage {
     }
   }
 
+  // TODO: async I don't know how to deal with that
```
Comment (saidelike) [inline packages/cursorless-engine/src/processTargets/modifiers/InstanceStage.ts:50]: @pokey do you know how to deal with that? it is a promise of iterator

image

File: packages/cursorless-engine/src/processTargets/targets/DestinationImpl.ts [hunk 3]
```diff
@@ -31,12 +33,21 @@ export class DestinationImpl implements Destination {
       indentationString ?? this.isLineDelimiter
         ? getIndentationString(target.editor, target.contentRange)
         : "";
-    this.insertionPrefix =
-      target.prefixRange != null
-        ? target.editor.document.getText(target.prefixRange)
+  }
+
+  // TODO: we can't have "async/await" in a constructor so we initialize elsewhere
```
Comment (saidelike) [inline packages/cursorless-engine/src/processTargets/targets/DestinationImpl.ts:38]: @pokey do we like this way to call async functions when constructing an object?

File: packages/cursorless-engine/src/processTargets/targets/InteriorTarget.ts [hunk 4]
```diff
@@ -12,20 +16,28 @@ export class InteriorTarget extends BaseTarget<InteriorTargetParameters> {
...
-  constructor(parameters: InteriorTargetParameters) {
-    super({
+  // TODO: don't use the constructor, instead we need to create the object with create()
+  // Are we happy about that?
+  constructor(parameters: InteriorTargetParameters & CommonTargetParameters) {
+    super(parameters);
+    this.fullInteriorRange = parameters.fullInteriorRange;
+  }
+
+  public static async create(
```
Comment (saidelike) [inline packages/cursorless-engine/src/processTargets/targets/InteriorTarget.ts:26]: @pokey or do we prefer this way to call async functions when constructing an object?

File: packages/cursorless-engine/src/processTargets/targets/DestinationImpl.ts (hunk 3 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/targets/DestinationImpl.ts:38]: I wonder if we could get the insertion prefix lazily when we actually need it?

File: packages/cursorless-engine/src/processTargets/targets/InteriorTarget.ts (hunk 4 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/targets/InteriorTarget.ts:26]: this is probably better, as it avoids client needing to call initializer. I'd also make the constructor private to force them to call `create`

File: packages/cursorless-engine/src/actions/BringMoveSwap.ts [hunk 5]
```diff
@@ -82,16 +82,16 @@ abstract class BringMoveSwap {
...
-            .map((source, i) => {
-              const text = source.contentText;
+            .map(async (source, i) => {
+              const text = await source.contentText;
               const delimiter =
                 (destination.isRaw ? null : destination.insertionDelimiter) ??
                 (source.isRaw ? null : source.insertionDelimiter);
               return i > 0 && delimiter != null ? delimiter + text : text;
             })
             .join("");
```
Comment (conartist6) [inline packages/cursorless-engine/src/actions/BringMoveSwap.ts:92]: Isn't this join happening on an array of promises now?

**Comment by pokey** [comment]
update from meet-up:

- Let's start by just grabbing the full buffer for active editor from neovim right before we execute the command. Not the most efficient thing, but seems like easiest way forward for now. We'll grab implementation from https://github.com/cursorless-dev/cursorless/pull/1899/files

**Comment by saidelike** [comment]
closing as over complicated and won't be needed after all
